infra/cifuzz/filestore/gitlab/__init__.py

```diff
@@ -0,0 +1,121 @@
+    """Gitlab artifacts implementation of upload_corpus."""
+    self._copy_fromdir(directory, name, "corpus")
+
+  def upload_build(self, name, directory):
+    """Gitlab artifacts implementation of upload_build."""
+    self._copy_fromdir(directory, name, "build")
+
+  def upload_coverage(self, name, directory):
+    """Gitlab artifacts implementation of upload_coverage."""
+    # would be neat if we could save CI_JOB_ID somewhere,
+    # without needing a token with write-access
+    self._copy_fromdir(directory, name, "coverage")
+
+  def _get_jobid(self, projpath, headers, reason):
+    """Get a specific job id for the latest succesful pipeline
+    with the specific job name."""
+    jobsurl = os.getenv('CI_API_V4_URL') + projpath + "/jobs?scope=success"
+    with tempfile.NamedTemporaryFile() as tmp_file:
+      if not http_utils.download_url(jobsurl, tmp_file.name, headers=headers):
+        logging.error('Failed downloading %s', jobsurl)
+        return None
+      # jq '.[] | select(.name=="clusterfuzzlite-corpus") | .id'
+      with open(tmp_file.name, encoding='ascii') as json_file:
+        data = json.load(json_file)
+        for j in data:
+          # job names are fixed by this
+          if j["name"] == "clusterfuzzlite-" + reason:
+            logging.info('Latest job with %s is %d', reason, j["id"])
+            return j["id"]
+    return None
+
+  def _copy_todir(self, dst, name, reason):
+    if reason not in self.downloaded:
+      branch = os.getenv('CFL_BRANCH')
+      if not branch:
+        branch = os.getenv('CI_DEFAULT_BRANCH')
+      projpath = "/projects/" + os.getenv(
+          'CI_PROJECT_NAMESPACE') + "%2F" + os.getenv('CI_PROJECT_NAME')
+      # headers = {"JOB-TOKEN" : os.getenv('CI_JOB_TOKEN')}
+      # is not enough to get jobid, we could just loop over them...
+      headers = {"PRIVATE-TOKEN": os.getenv('CFL_PRIVATE_TOKEN')}
+      jobid = self._get_jobid(projpath, headers, reason)
+      if not jobid:
+        logging.error('Could not find job id for %s', reason)
+      else:
+        srcurl = os.getenv('CI_API_V4_URL') + projpath + "/jobs/" + str(
+            jobid) + "/artifacts"
+        logging.info('Downloading artifacts from %s', srcurl)
+        dldir = os.path.join(self.config.workspace, self.download_dir)
+        os.makedirs(dldir, exist_ok=True)
+        http_utils.download_and_unpack_zip(srcurl, dldir, headers=headers)
+        self.downloaded.add(reason)
+    if reason in self.downloaded:
+      srcdir = os.path.join(self.config.workspace, self.download_dir,
+                            self.artifacts_dir, reason, name)
+      logging.info('Downloading %s to artifacts to %s', srcdir, dst)
+      shutil.copytree(srcdir, dst, dirs_exist_ok=True)
+      return True
+    return False
+
+  def download_corpus(self, name, dst_directory):
+    """Gitlab artifacts implementation of download_corpus."""
+    self._copy_todir(dst_directory, name, "corpus")
```

Comment:
nit: single quotes and for rest of this file.

Reply:
ok